fix: Label handling

embano1 opened the pull request:
Follow-up on #194 where the logic was still not correct in all cases. The behavior is now as expected: if `--labels` is specified all keys and values specified must match a feature. This allows for arbitrary and complex combinations of multiple features within one suite.

For example:

```go
f1 := features.New("label demo 1").
		WithLabel("a", "1").
		WithLabel("all", "true").
		Feature()

f2 := features.New("label demo 2").
		WithLabel("all", "true").
		Feature()

f3 := features.New("label demo 3").
		WithLabel("a", "2").
		WithLabel("all", "true").
		Feature()

f4 := features.New("label demo 4").
		WithLabel("a", "1").
		WithLabel("b", "1").
		WithLabel("all", "true").
		Feature()
```

Leads to the following outcomes with different `--labels`:

```console
go test -race -v ./e2e -args -v 4 --labels=some=value
--- PASS: TestDemoE2E (0.07s)
    --- SKIP: TestDemoE2E/label_demo_1 (0.00s)
    --- SKIP: TestDemoE2E/label_demo_2 (0.00s)
    --- SKIP: TestDemoE2E/label_demo_3 (0.00s)
    --- SKIP: TestDemoE2E/label_demo_4 (0.00s)
PASS
```

```console
go test -race -v ./e2e -args -v 4 --labels=a=1
--- PASS: TestDemoE2E (0.06s)
    --- PASS: TestDemoE2E/label_demo_1 (0.00s)
    --- SKIP: TestDemoE2E/label_demo_2 (0.00s)
    --- SKIP: TestDemoE2E/label_demo_3 (0.00s)
    --- PASS: TestDemoE2E/label_demo_4 (0.00s)
PASS
```

```console
go test -race -v ./e2e -args -v 4 --labels=all=true
--- PASS: TestDemoE2E (0.05s)
    --- PASS: TestDemoE2E/label_demo_1 (0.00s)
    --- PASS: TestDemoE2E/label_demo_2 (0.00s)
    --- PASS: TestDemoE2E/label_demo_3 (0.00s)
    --- PASS: TestDemoE2E/label_demo_4 (0.00s)
PASS
```

```console
go test -race -v ./e2e -args -v 4 --labels=all=true,a=1
--- PASS: TestDemoE2E (0.04s)
    --- PASS: TestDemoE2E/label_demo_1 (0.00s)
    --- SKIP: TestDemoE2E/label_demo_2 (0.00s)
    --- SKIP: TestDemoE2E/label_demo_3 (0.00s)
    --- PASS: TestDemoE2E/label_demo_4 (0.00s)
PASS
```

Passing no `--labels` of course works as expected

```console
go test -race -v ./e2e -args -v 4
--- PASS: TestDemoE2E (0.06s)
    --- PASS: TestDemoE2E/label_demo_1 (0.00s)
    --- PASS: TestDemoE2E/label_demo_2 (0.00s)
    --- PASS: TestDemoE2E/label_demo_3 (0.00s)
    --- PASS: TestDemoE2E/label_demo_4 (0.00s)
PASS
```

Also works in combination with `--skip-labels`

```console
go test -race -v ./e2e -args -v 4 --labels=a=1 --skip-labels=all=true
--- PASS: TestDemoE2E (0.05s)
    --- SKIP: TestDemoE2E/label_demo_1 (0.00s)
    --- SKIP: TestDemoE2E/label_demo_2 (0.00s)
    --- SKIP: TestDemoE2E/label_demo_3 (0.00s)
    --- SKIP: TestDemoE2E/label_demo_4 (0.00s)
PASS
```

vladimirvivien reviewed (commented):
@embano1 This looks great. Thanks for the awesome work.

I will ask for one favor though:
Can you update examples/skip_flags/README.md with the examples you have on the PR that shows the complex combination that is now supported!

embano1 commented:
@vladimirvivien done! Formatted and added more details to the README for clarity. Also updated the test examples to show how multiple labels can be used.

ShwethaKumbla commented:
/ok-to-test

vladimirvivien reviewed (commented):
Thanks a lot for this and the doc update.
Leaving it as `comment` for now so you can see my comment on the review.

/lgtm

embano1 commented on examples/skip_flags/README.md:
```diff
@@ -20,22 +23,17 @@ func TestMain(m *testing.M) {
-### Supported flags
+### Use `Labels` in your tests
 
-There are several supported flags (for more accurate list, see package `pkg/flag`):
```
The reason I removed it as we have this already covered in the `flags` [example](https://github.com/kubernetes-sigs/e2e-framework/tree/main/examples/flags#supported-flags) and this one is focused on `labels`. But will add tracking issue.

vladimirvivien replied:
Ok, if already documented, no need for the tracking issue.

embano1 commented:
@vladimirvivien tide is still pending?

vladimirvivien commented:
/approve
